akka-projection-core/src/main/scala/akka/projection/Projection.scala

```diff
@@ -60,40 +67,36 @@ trait Projection[Envelope] {
 @InternalApi
 private[projection] object RunningProjection {
 
-  def withBackoff(source: () => Source[Done, _], settings: ProjectionSettings): Source[Done, _] =
+  def withBackoff(source: () => Source[Done, _], settings: ProjectionSettings): Source[Done, _] = {
     RestartSource
       .onFailuresWithBackoff(settings.minBackoff, settings.maxBackoff, settings.randomFactor, settings.maxRestarts) {
         () => source()
       }
-
-  /* internal exception to wrap exceptions coming from stopHandler */
-  private case class StopHandlerException(cause: Throwable) extends RuntimeException(cause) with NoStackTrace
+  }
 
   /**
    * Adds a `watchTermination` on the passed Source that will call the `stopHandler` on completion.
    *
    * The stopHandler function is called on success or failure. In case of failure, the original failure is preserved.
    */
-  def stopHandlerOnTermination(src: Source[Done, NotUsed], stopHandler: () => Future[Done])(
-      implicit ec: ExecutionContext): Source[Done, Future[Done]] = {
+  def stopHandlerOnTermination(
```

Comment:
@renatocaval one more. There was a race between the order of restarted and stopped, found by Travis. Fixed by doing both here. Also simplified by using `andThen`. See full fix of that in 94939a1